src/magic_folder/util/database.py

```diff
@@ -0,0 +1,132 @@
+# Copyright 2020 The Magic-Folder Developers
+# See COPYING for details.
+
+"""
+Utilties for dealing with sqlite.
+"""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import contextlib
+import inspect
+import sqlite3
+
+import attr
+from six import wraps
+from twisted.python.compat import currentframe
+
+__all__ = [
+    "LockableDatabase",
+    "RecusiveTransaction",
+    "WithCursorGenerator",
+    "with_cursor",
+]
+
+
+@attr.s(auto_exc=True, frozen=True)
+class WithCursorGenerator(TypeError):
+    """
+    :py:`with_cursor` cannot be used on a generator.
+
+    As it is currently implemented, the transaction would be commited before
+    any code in the decorated generator ran. Even if we special cased generators,
+    having the transaction stay open until the generator completes, that loses
+    the proprerty of `with_cursor` being transparent and atomic from the point
+    of view of the calller.
+    """
+
+    function = attr.ib()
+
+    def __str__(self):
+        return "'with_cursor' cannot decorate a generator function: {!r}".format(
+            self.function
+        )
+
+
+@attr.s(auto_exc=True, frozen=True)
+class RecusiveTransaction(Exception):
+    function = attr.ib(default=None)
+    caller = attr.ib(default=None)
+
+    def __str__(self):
+        message = "Tried to enter a recursive transaction"
+        if self.function:
+            message += " when calling '{}'".format(self.function)
+        if self.caller:
+            message += " from '{}'".format(self.caller)
+        message += "."
+        return message
+
+
+@attr.s
+class LockableDatabase(object):
+    """
+    Wrapper around an :py:`sqlite3.Connection`
+    """
+
+    _database = attr.ib(validator=attr.validators.instance_of(sqlite3.Connection))
+    _in_transaction = attr.ib(
+        default=False, init=False, validator=attr.validators.instance_of(bool)
+    )
+
+    @contextlib.contextmanager
+    def transaction(self, function_name=None):
+        """
+        Context manager that starts a transaction and rollsback
+        if an exception is raised.
+
+        :param Optional[function_name]: The name of the calling function, used
+            in error messages.
+
+        :raises RecusiveTransaction: if used recusively.
+
+        :return sqlite3.Cursor:
+        """
+        if self._in_transaction:
+            frame = currentframe(2)
+            if frame.f_code == with_cursor._wrapped_code:
+                frame = frame.f_back
+            caller = frame.f_code.co_name
+            raise RecusiveTransaction(function=function_name, caller=caller)
```

Comment:
And I guess I shouldn't say anything about `SAVEPOINT` here... maybe some other time.

Reply:
> And I guess I shouldn't say anything about SAVEPOINT here... maybe some other time.

I did spend a bit of time looking at that. I think we'd need to switch to managing transactions manually, rather than using pysqlite's partially automatic handling (see [here](https://docs.python.org/2.7/library/sqlite3.html#controlling-transactions)). In particular, we'd need to manually handle commit vs rollback as `with database` doesn't understand `SAVEPOINT` (and I had thought that it didn't work with regular transaction in manual mode, but reading the py2.7 code, it does still seem to work for that).

It may be worth doing that in the future (at the least stopping using the automatic transaction management).

Reply:
> Also, the full stack is necessary (or at least desirable) to debug the problem in the general case. It is just as much an error to attempt to begin a nested transaction after going through a function not decorated with with_cursor - so currentframe(2) doesn't necessarily get you to anything particularly interesting, database-wise.

The `2` is about the `contextlib.contextmanager` wrapper (specifically `GeneratorContextManager.__enter__` I think). So 
`currentframe(2)` should always point at the caller of `.transaction()`. I've got a commit with a comment explaining that I had written awhile ago adding a `.close` for https://github.com/LeastAuthority/magic-folder/pull/475/commits/f4bbe95402dbddb2b052e77ab29042d2ad4b85ca (which I only discovered I had needed after opening this PR). I'll push that here, since otherwise this will fail once its merged.

> Eh ... I am not excited to see stack and frame inspection code here.

I don't feel strongly about having this, though I do think it is valuable to include the *called* function name for the `with_cursor` case, since that won't be in the stack trace. I don't recall exactly why I added it; it may have been for convenience when figuring out where we were triggering this, or it may been so the message would have symmetry of "from <caller>" vs the "when calling <callee>".

I suspect part of my thinking was that this message is probably enough to diagnose the issue *without* needing the full stack trace, so spending a bit of effort here on a good error message will save time down the road (particularly given our current usage of transactions in `@property` everywhere).

Reply:
> I did spend a bit of time looking at that. I think we'd need to switch to managing transactions manually, rather than using pysqlite's partially automatic handling (see here).

For what it's worth, I don't think there was any particular *intent* to use pysqlite's partially automatic transaction handling - it's just what you get if you don't read a bunch of extra documentation and carefully choose your connect parameters.  So, yes, at some point it probably makes sense to move away from that.